### PR TITLE
WIP: Shipley/app_directory_endpoint

### PR DESCRIPTION
**Pull request by bradenshipley**
## Description of change
In order to expose what apps are available to end users and to populate the App Directory currently in development, I've added the apps_api module. This module contains a single controller with one route (`/apps`) to expose all oauth apps in our okta organization. The results are filtered to remove api_tokens listed as applications and cached in redis. There is a scheduled rake task every 8 hours to update the cache. 

## Original issue(s)
https://vajira.max.gov/browse/API-1236

## Things to know about this PR
This PR contains many files changed, which is a necessary evil due to all the boilerplate needed.
Updates to config/routes.rb and sidekiq scheduler to include new module. New service has been created and existing okta service has been modified. Everything else is contained in the new module.
In `lib/okta/directory_service.rb` rubocop is disabled to account for the long regex that matches any ISO date. If this is an issue it could be refactored into multiple smaller regex patterns that are joined together.
Basic documentation for this endpoint is included in the `modules/apps_api/app/swagger` directory.

Basic Rspec tests for the endpoint using VCRs are implemented.

### CODE REVIEW

File: lib/okta/directory_service.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      unfiltered_apps
+    end
+
+    # Check if headers contains 'next' link, since 'next' is always after 'self', we can use .last as a consistent check
+    def contains_next(headers)
+      headers['link'].split(',').last.include?('next')
+    end
+
+    def substring_next_link(headers)
+      headers['link'].split(',').last.split('<').last.split('>').first
```
Comment (bastosmichael) [inline lib/okta/directory_service.rb:52]: Include some &. to the front of those methods.

File: lib/okta/directory_service.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  class DirectoryService < Common::Client::Base
+    def get_apps
+      okta_service = Okta::Service.new
+      redis = Redis.new
+      base_url = Settings.oidc.base_api_url + '/api/v1/apps?limit=200&filter=status+eq+\"ACTIVE\"'
+      unfiltered_apps = recursively_get_apps(okta_service, base_url)
+
+      # matches any ISO date.
+      # rubocop:disable Metrics/LineLength
+      iso_pattern = /(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+)|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d)|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d)/
```
Comment (bastosmichael) [inline lib/okta/directory_service.rb:15]: Can this be a constant instead? ISO_PATTERN? and do a .freeze at the end?

**Review by bastosmichael — DISMISSED** [review]
Make sure to include proper swagger documentation for the portal as well as possible pagination in case this data set from Okta ever blows up to an unmanageable number of apps. Otherwise looks good minus a few other comments I already left.

File: lib/okta/directory_service.rb (hunk 2 above)
Comment (bradenshipley) [reply lib/okta/directory_service.rb]: Done!

File: lib/okta/directory_service.rb (hunk 1 above)
Comment (bradenshipley) [reply lib/okta/directory_service.rb]: Done!

**Review by bradenshipley — COMMENTED** [review]
Swagger documentation is included in the 4b466c6 commit

**Review by annaswims — CHANGES_REQUESTED** [review]
Please revert your changes to sidekiq in Gemfile.lock

File: config/database.yml [hunk 3]
```diff
@@ -28,10 +28,10 @@ development:
   # To create additional roles in postgres see `$ createuser --help`.
   # When left blank, postgres will use the default role. This is
   # the same name as the operating system user that initialized the database.
-  #username: postgres
```
Comment (annaswims) [inline config/database.yml:31]: revert this too

File: modules/apps_api/Gemfile.lock [hunk 4]
```diff
@@ -0,0 +1,111 @@
+PATH
```
Comment (annaswims) [inline modules/apps_api/Gemfile.lock:1]: I don't think we want to commit a Gemfile.lock for a module

File: vcr.log [hunk 5]
```diff
@@ -0,0 +1,152 @@
+[Cassette: 'okta/directories'] Initialized with options: {:record=>:once, :record_on_error=>true, :match_requests_on=>[:method, :uri], :allow_unused_http_interactions=>true, :serialize_with=>:yaml, :persist_with=>:file_system, :persister_options=>{}}
```
Comment (annaswims) [inline vcr.log:1]: Why was this committed?

File: config/database.yml (hunk 3 above)
Comment (bradenshipley) [reply config/database.yml:31]: Done!

File: vcr.log (hunk 5 above)
Comment (bradenshipley) [reply vcr.log:1]: Purely an error.

File: modules/apps_api/Gemfile.lock (hunk 4 above)
Comment (bradenshipley) [reply modules/apps_api/Gemfile.lock]: This file was unintentionally added. Done!

**Comment by bradenshipley** [comment]
> Please revert your changes to sidekiq in Gemfile.lock

Done! This file was unintentionally added and has been removed.

File: modules/apps_api/app/swagger/apps_api/v0/apps.yml [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    This API is available publically without an API Key.
+
+    1. Client Request: GET https://sandbox-api.va.gov/services/apps/v0/directory
+
+    2. Service Response: A JSON API object with a list of all applications available
+
+
+  termsOfService: 'https://developer.va.gov/terms-of-service'
+  contact:
+    name: Vets.gov
```
Comment (charleystran) [inline modules/apps_api/app/swagger/apps_api/v0/apps.yml:21]: This should be va.gov not vets.gov. We probably havent updated it in other modules yet.

File: modules/apps_api/config/locales/en.yml [hunk 7]
```diff
@@ -0,0 +1,26 @@
+en:
+  va_forms:
+    endpoint_descriptions:
+      index: 'A listing of available VA forms and their location.'
```
[the remark this answers is not in the document]
Comment (bradenshipley) [reply modules/apps_api/config/locales/en.yml]: Fixed in commit # e9d4780

**Comment by bradenshipley** [comment]
Closing due to change in direction

**Pull request closed without merging**